### ISSUE TITLE
Variables in JSX strings put on new line

### ISSUE THREAD

**Issue by thchia** [kentcdodds/react-testing-library#53]
* `react-testing-library` version: 2.1.0
* `node` version: 8.9.3
* `npm` (or `yarn`) version: `yarn` 1.3.2

Relevant code or config

Component to be tested:
```javascript
const Component = ({ currentStep }) => <div>Step {currentStep} of 4</div>
```

What you did:

Test code:
```javascript
const { getByText } = render(<Component currentStep={1} />)

expect(getByText('Step 1 of 4')).toBeDefined()
```

What happened:

```javascript
Unable to find an element with the text: Step 1 of 4. This could be because the text is broken up by multiple elements. In this case, you can provide a function for your text matcher to make your matcher more flexible.

<div>
    Step
    1 // <--------- newline
      of 4
 </div>
```

Reproduction repository:

This can be reproduced in the current repository test cases by inserting a variable into the JSX.

Problem description:

Variables in JSX strings cause new line issues (and failing tests). I believe that in the spirit of "not testing implementation details", this should be considered a harmful problem. If I add an extra space around the "1", the test passes (period signifies space):

```javascript
expect(getByText('Step..1..of 4')).toBeDefined() // passing
```

Suggested solution:

I don't know why this is happening, but I'm guessing it is something to do with how the component is being serialised. I'm also guessing that this issue should be here and not dom-testing-library but I don't know for sure where the root of the problem lies.

**Comment by gnapse** [comment]
Given that in html any extra whitespace between words in text is not meaningful, and the resulting text shown to the user on screen is what you expect in your test above, I think it makes sense to take this into account when matching an element's text content.

Actually I am now wondering if `.toHaveTextContent` should also take this into account.

**Comment by kentcdodds** [comment]
Yes. We should update the `matcher` function in dom-testing-library to replace multiple spaces with one before it attempts to do any comparison. I consider this a bug and would love a pull request for it 👍

**Comment by gnapse** [comment]
I'll take it, since I'll probably also have to do this to jest-dom's `.toHaveTextContent` ;)

**Comment by DarrylD** [comment]
I'm still having this issue, any idea why it would continue to add white spaces? Updated to the latest dom-testing-library.

**Comment by gnapse** [comment]
@DarrylD your best bet for others to be able to help is to provide a sample repository or codesandbox where the problem can be reproduced.

**Comment by max-sr** [comment]
I have also continued to experience the same issue, and was able to diagnose it (as far as my case goes, at least).

The problem is that the amendment to `getText` presupposes that when text is divided over multiple lines, the text on each line will constitute a separate word. Things go wrong when this is not the case; for example:

```
Pay
£
24.99
```

will end up as `Pay £ 24.99`. This problem is remedied if [line 73 here](https://github.com/gnapse/dom-testing-library/blob/a6a9d2c10a091aa1349a840d73602fef8723358c/src/queries.js#L73) is changed from `.join(' ')` to `.join('')`.

Then, however, you'd be required to ensure a space is present after each separate word that's on a different line, so `Pay` would have to be `Pay ` (i.e. with a space after it on the same line) in my example. Otherwise, there's no way to differentiate between words on different lines that should be separated by spaces on the one hand, and word fragments/characters on separate lines that should be concatenated on the other.

But I'm not really proposing this change be made, as you shouldn't have to ensure those spaces are present in your JSX.

**Comment by maxcct** [comment]
I have also continued to experience the same issue, and was able to diagnose it (as far as my case goes, at least).

The problem is that the amendment to `getText` presupposes that when text is divided over multiple lines, the text on each line will constitute a separate word. Things go wrong when this is not the case; for example:

```
Pay
£
24.99
```

will end up as `Pay £ 24.99`. This problem is remedied if [line 73 here](https://github.com/gnapse/dom-testing-library/blob/a6a9d2c10a091aa1349a840d73602fef8723358c/src/queries.js#L73) is changed from `.join(' ')` to `.join('')`.

Then, however, you'd be required to ensure a space is present after each separate word that's on a different line, so `Pay` would have to be `Pay ` (i.e. with a space after it on the same line) in my example. Otherwise, there's no way to differentiate between words on different lines that should be separated by spaces on the one hand, and word fragments/characters on separate lines that should be concatenated on the other.

But I'm not really proposing this change be made, as you shouldn't have to ensure those spaces are present in your JSX.

**Comment by kentcdodds** [comment]
This is actually how HTML works. If you were to put this in the DOM:

```javascript
<div>
Pay
£
24.99
</div>
```

The resulting output would appear to the user as:

`Pay £ 24.99`

This is why we do the `join(' ')` in dom-testing-library.

**Comment by maxcct** [comment]
That is a very good point. I guess then all you can do is ensure that you concatenate strings like this before inserting them into JSX.

**Comment by bdchauvette** [comment]
I'm running into the same problem, even with strings that don't have a space between the variable and normal text, e.g. `<span>{volts}V<span>`.

In this situation, the DOM node's `textContent` property contains a string without spaces, even though there are two `Text` nodes behind the scenes.

Live example: https://codesandbox.io/s/w718n5ojq7

**Comment by kentcdodds** [comment]
Ah, this is tricky! What do you propose?

**Comment by kentcdodds** [comment]
Yeah, this should be talked about in `dom-testing-library`. Could you make a PR there and we can discuss how to solve this there. Thanks!

**Comment by anapanadero** [comment]
I continue with the same problem. For example:
`<span>My name is (${name})</span>` 
The result for this code is:
`My name is (
Ana
)`
So getByText will fail saying _Unable to find an element with the text_ when I try to do 
`screen.getByText('My name is ${name})')`

What can I do to make this work?

**Comment by anapanadero** [comment]
I continue with the same problem. For example:
```<span>My name is (${name})</span>```
The result for this code is:
```
<span> 
My name is ( 
 Ana
)
</span>
```
So getByText will fail saying _Unable to find an element with the text_ when I try to do 
```screen.getByText(`My name is ${name})`)```

What can I do to make this work?

**Comment by tubbo** [comment]
This is also happening for me.

My JSX:

```typescriptreact
          <Text as="h1" size="large">
            Leave {data.community.name}?
          </Text>
```

The output:

```html
            <h1
              class="c-PJLV c-PJLV-ddalwR-size-large"
            >
              Leave
              Foo
              ?
            </h1>
```

And the test:

```typescript
  it('confirms with the user whether they wish to leave', async () => {
    expect(
      await screen.findByText('You are about to leave the Foo community'),
    ).toBeInTheDocument()
  })
```

Not sure if this is actually the same issue, but it's definitely the same symptom. Noticed it when I upgraded to React 17 and started using the new JSX transform.

**Comment by klauskpm** [comment]
> I was able to work around this by changing my assertion to:
> 
> ```ts
> expect(
>   await screen.findByText(/You are about to leave the(.*)Foo(.*)community./)
> ).toBeInTheDocument()
> ```

I'm using an older version, so maybe an update (not an option right now) would solve this, but I got it working with a solution similar to yours:
```js
expect(screen.getByRole('button', { name: /Factors\s/ })).toBeInTheDocument();
```

**Comment by OttlikG** [comment]
This is still an issue for me. Is there going to be any solution for this?

**Comment by alexander-schranz** [comment]
Is there any issue about changing the behaviour how this is rendered and so snapshots for me it is just looks false if I have component like this:

```jsx
<div>
    {charactersLeft} {translate('sulu_admin.characters_left')}
<div>
```

Is snapshoted as:

```html
<div>
    -8
     
    sulu_admin.characters_left
</div>
```

Instead of:

```html
<div>
    -8 sulu_admin.characters_left
</div>
```

**Comment by b3h3m0th** [comment]
We managed to solve this problem by manually joining the strings like so:

```jsx
<div>
    {charactersLeft + ' ' + translate('sulu_admin.characters_left')}
</div>
```

The above generates the following (desired) snapshot:

```jsx
<div>
    -8 sulu_admin.characters_left
</div>
```

**Comment by evanjmg** [comment]
getting something very similar here - very strange behaviour. This issue is not closed. 
  - Snapshot  - 0
    + Received  + 1

    @@ -1,6 +1,7 @@
      <DocumentFragment>
    +    
        <div
          class="page-heading"
        >

it's adding new lines above my elements

**Comment by chris-hinds-cko** [comment]
This is still an issue in 2023, None of the above solutions seem to work for me.

rendering `Hello {name}, How are you`

still results in

```
<h1
            class="Heading__TextHeadingOne-sc-1u68qyt-0 kLnbPE"
            id="heading"
          >
            Hello
            Expedia,
             How are you
          </h1>
```

**Comment by pstevovski** [comment]
Any workaround for the above mentioned issues? This is a really old issue by now, and no idea why its marked as closed, when no specific answer was provided (by anyone) on how should we handle such scenarios.

I, as many others have already stated, am having an issue where the following JSX in my component:

```js
<div>
  <h2>{status} text ({count})</h2>
</div>
```

is rendered like this in the tests:
```js
<h2>
new
ads (
1
 )
 </h2>
```

And I cannot use `getByText` matcher to target it, as it says its unable to find the element. So.. how can I target it? I do not want to include `data-testid` attribute in this case (as this is part of a dynamically generated content for a table).

**Comment by alexander-schranz** [comment]
@pstevovski workaround by @b3h3m0th should work in your case: https://github.com/testing-library/react-testing-library/issues/53#issuecomment-1198173548

**Comment by pstevovski** [comment]
@alexander-schranz Thank you for your answer, and sorry for the late reply, no idea why but I didn't get any notification.

The answer provided by @b3h3m0th doesn't really work for my case in terms of practicality, I'd have to go and change a lot of the pieces trough out the application to convert them to be manually concatenated like that. And also, to be fair, that is just a "workaround" (as you mentioned), but then again accessing and reading dynamic values in React (or in JS in general) has better ways like we're all already using (either that be template literals or the way we use them in react).

Anyhow, here's a snippet that I found, with small modifications, which does a lookup based on the markup:
`

type Query = (queryMatcher: MatcherFunction) => HTMLElement;

/**
 *
 * Helper matcher function to be used together with React Testing Library matchers.
 *
 * Helps find an element in the rendered testing UI that has nested HTML elements:
 *
 * @example
 *  <h1>
 *    <span>Title</span> with
 *    <strong>nested</strong> elements.
 *  </h1>
 *
 * @param query The query matcher function to be used, provided by RTL
 * @param text The text by which we want to find the targeted element
 *
 * @returns The targeted element (if found). Otherwise, throws an error.
 *
 */
export const getByTextWithMarkup = (query: Query, text: string): HTMLElement => {
  return query((_content: string, node: any) => {
    const hasText = (node: HTMLElement) => node.textContent === text;
    const childrenDontHaveText = Array.from(node.children).every(
      child => !hasText(child as HTMLElement),
    );
    return hasText(node) && childrenDontHaveText;
  });
};
`
This worked just fine for me.

I know this is a closed topic, but:

1. This might be useful to someone out there
2. I think that there should definitely be some sort of 'in-built' way for RTL to handle these scenarios.

**Comment by ErikBrendel** [comment]
I found a workaround: Apply Node.normalize before taking a snapshot → this merges adjacent nodes into one. See https://developer.mozilla.org/en-US/docs/Web/API/Node/normalize

Inside e.g. your vitest setup file, add something like this:

```ts
// Normalize DOM nodes before snapshotting to merge adjacent text nodes
const normalizedNodes = new WeakSet<Node>()
expect.addSnapshotSerializer({
  test: (val) => val instanceof Node && !normalizedNodes.has(val),
  serialize: (val: Node, config, indentation, depth, refs, printer) => {
    normalizedNodes.add(val)
    val.normalize()
    const result = printer(val, config, indentation, depth, refs)
    normalizedNodes.delete(val)
    return result
  },
})
```

Could this maybe be adopted as testing-library core functionality? Maybe opt-in to not break existing snapshots?
